### ISSUE TITLE
Question: Webpack-style hmr hook

### ISSUE THREAD

**Issue by natew** [steelbrain/pundle#173]
Wondering how I'd do this with 3.0.

Previously I remember we had some compile time thing that passed `module` into our `@view` decorator at certain places so it would accept all over.

Now, I have it more standard. Just react-hot-loader latest and it accepts at the root. In the single file below that accept, I have this:

![image](https://user-images.githubusercontent.com/12100/42475629-07e2e15a-8380-11e8-835b-7cc3ed9be40b.png)

This basically saves all the states, and then reloads them after HMR.

Basically, wondering if I can hook into pre/post HMR. It may be worth adding this style hook just to be compat.

Secondarily, would be interested in avoiding full-tree reloads, though if it's only calling the handler and not actually re-compiling the tree that may be unnecessary complexity.

**Comment by steelbrain** [comment]
Pundle does not subscribe/check for status changes, It applies as soon as it gets it

According to the [HMR API](https://webpack.js.org/api/hot-module-replacement/) there's a `addDisposeHandler` that you can add and emit `will-hmr` in, and then do `module.hot.accept('*', callback)` in Pundle or `module.hot.accept('.', callback)` in Webpack and in that callback you can emit the `did-hmr` as it's invoked after the module is replaced

This is not ideal though, and will vary from bundler to bundler. (Pundle may not support `['*']` in the future).

I think the least side effective way to do it is to do something like

```js
// Added process env check so static analysis can remove this node in prod
// instead of ignoring on runtime
if (process.env.NODE_ENV === 'development' && module.hot) {
  const { data } = module.hot
  if (data.oldView) {
    data.oldView.emit('did-hmr')
  }
  module.hot.addDisposeHandler(function(newData) {
    view.emit('will-hmr')
    newData.oldView = view
  })
}
```

**Comment by RaeesBhatti** [comment]
This is fixed by https://github.com/steelbrain/pundle/pull/175
